### DISCUSSION TITLE
CustomPartProps needs to be added to payloads and attachments for WSS4JAttachment when using the `Phase4XXXSender.Builder`

### DISCUSSION THREAD

**Discussion by oaxfreeman** [phax/phase4#276] [Q&A] [answered]
I need this class:
![image](https://github.com/user-attachments/assets/61593c28-2788-4c6e-b1b6-b23bfb36654c)

to reflect this XML structure:
![image](https://github.com/user-attachments/assets/af50a127-f7f5-4fcd-8234-09db2818a89d)

Builder doesn't seem to allow for extra property addition.

@phax Is there any other mechanism to customize outgoing envelope?

**Comment by phax** [comment]
Agree. The class for the outgoing attachment needs this property.
I assume you need it for the 2.x branch?

**Comment by oaxfreeman** [comment]
@phax would be nice..  so far I have managed to fork `AbstractAS4UserMessageBuilderMIMEPayload`

and override:
```java
  @Nullable
  @OverrideOnDemand
  protected WSS4JAttachment createMainAttachment (@Nonnull final AS4OutgoingAttachment aPayload,
                                                  @Nonnull final AS4ResourceHelper aResHelper) throws IOException
  {
    WSS4JAttachment attachment = WSS4JAttachment.createOutgoingFileAttachment (aPayload, aResHelper);
    attachment.customPartProperties().addAll(((MockAS4OutgoingAttachment) aPayload).getProperties());
    return attachment;
  }

  @Nullable
  @OverrideOnDemand
  protected WSS4JAttachment createOtherAttachment (@Nonnull final AS4OutgoingAttachment aPayload,
                                                   @Nonnull final AS4ResourceHelper aResHelper,
                                                   @Nonnegative final int nAttachmentIndex) throws IOException
  {
    WSS4JAttachment attachment = WSS4JAttachment.createOutgoingFileAttachment (aPayload, aResHelper);
    attachment.customPartProperties().addAll(((MockAS4OutgoingAttachment) aPayload).getProperties());
    return attachment;
  }
  ```
  While `MockAS4OutgoingAttachment` is just extending `AS4OutgoingAttachment ` with `Map<string, string>`

**Reply by phax** [reply, marked as answer]
Thanks, will do a fix for the 2.x branch to add custom properties as well

**Reply by phax** [reply]
Will be part of the 2.8.6 release and 3.0.0-beta2
